src/dbnode/storage/series/reader.go

```diff
@@ -119,88 +119,174 @@ func (r *Reader) readersWithBlocksMapAndBuffer(
+			i.curr = append(i.curr, i.buffer[0][0])
+			i.buffer[0] = i.buffer[0][1:]
+			if len(i.buffer[0]) == 0 {
+				i.buffer = i.buffer[1:]
+			}
+		}
+
+		// next check for the disk block at this time.
+		// check the cache first.
+		if len(i.cached) > 0 && !i.cached[0].Start.After(i.blockAt) {
+			i.curr = append(i.curr, i.cached[0])
+			i.cached = i.cached[1:]
+		} else {
+			// if not in the cache, request a load from disk.
+			blockReader, found, err := i.reader.streamBlock(ctx, i.blockAt, i.reader.onRetrieve, i.nsCtx)
+			if err != nil {
+				i.err = err
+				return false
+			}
+			if found {
+				i.curr = append(i.curr, blockReader)
+			}
+		}
+		i.blockAt = i.blockAt.Add(i.blockSize)
+	}
+	return len(i.curr) != 0
+}
+
+func (i *blockReaderIter) ToSlices(ctx context.Context) ([][]xio.BlockReader, error) {
+	var results [][]xio.BlockReader
+	for i.Next(ctx) {
+		results = append(results, i.Current())
+	}
+	if i.Err() != nil {
+		return nil, i.Err()
+	}
+	return results, nil
+}
+
 func (r *Reader) readersWithBlocksMapAndBufferAligned(
 	ctx context.Context,
 	start, end time.Time,
 	seriesBlocks block.DatabaseSeriesBlocks,
 	seriesBuffer databaseBuffer,
 	nsCtx namespace.Context,
-) ([][]xio.BlockReader, error) {
+) (BlockReaderIter, error) {
 	var (
 		nowFn       = r.opts.ClockOptions().NowFn()
 		now         = nowFn()
 		ropts       = r.opts.RetentionOptions()
 		blockSize   = ropts.BlockSize()
 		readerCount = end.Sub(start) / blockSize
+		buffer      [][]xio.BlockReader
+		cached      []xio.BlockReader
 	)
 
 	if readerCount < 0 {
 		readerCount = 0
 	}
 
-	// Two-dimensional slice such that the first dimension is unique by blockstart
-	// and the second dimension is blocks of data for that blockstart (not necessarily
-	// in chronological order).
-	//
-	// ex. (querying 2P.M -> 6P.M with a 2-hour blocksize):
-	// [][]xio.BlockReader{
-	//   {block0, block1, block2}, // <- 2P.M
-	//   {block0, block1}, // <-4P.M
-	// }
-	results := make([][]xio.BlockReader, 0, readerCount)
 	for blockAt := start; !blockAt.After(end); blockAt = blockAt.Add(blockSize) {
-		// resultsBlock holds the results from one block. The flow is:
-		// 1) Look in the cache for metrics for a block.
-		// 2) If there is nothing in the cache, try getting metrics from disk.
-		// 3) Regardless of (1) or (2), look for metrics in the series buffer.
+		// Eagerly load the readers from the series buffer and disk cache to avoid holding locks in the iterator.
 		//
 		// It is important to look for data in the series buffer one block at
 		// a time within this loop so that the returned results contain data
 		// from blocks in chronological order. Failure to do this will result
 		// in an out of order error in the MultiReaderIterator on query.
-		var resultsBlock []xio.BlockReader
 
-		blockReader, block, found, err := retrieveCached(ctx, blockAt, seriesBlocks)
+		blockReader, blk, found, err := retrieveCached(ctx, blockAt, seriesBlocks)
```

Comment:
we still eagerly read the cache and series buffer since they are already in memory. avoids having to holds locks while iterating.